<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Anne Senabouth" />

<meta name="date" content="2018-08-08" />

<title>‘ascend’: Installation and environment setup</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; position: absolute; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; }
pre.numberSource a.sourceLine:empty
  { position: absolute; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: absolute; left: -5em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>

</head>

<body>




<h1 class="title toc-ignore">‘ascend’: Installation and environment setup</h1>
<h4 class="author"><em>Anne Senabouth</em></h4>
<h4 class="date"><em>2018-08-08</em></h4>



<div id="before-you-begin" class="section level2">
<h2>Before you begin</h2>
<div id="system-requirements" class="section level3">
<h3>System requirements</h3>
<p>Datasets produced by single cell RNA sequencing (scRNA-seq) experiments are very large and can range from a few hundred to a million cells. The number of cells affect the amount of computational resources required to process the dataset – therefore, you need to determine if you have enough computational power and time to complete the analysis. <code>ascend</code> can comfortably analyse datasets of up to 10,000 cells on a single machine with 8GB of RAM and a quad-core CPU. Larger datasets should be run on a High Performance Cluster (HPC).</p>
<p>We have tested this package on datasets ranging from 100 to 20,000 cells. Generally, increasing the number of CPUs will decrease the processing time of functions, while larger datasets require more RAM.</p>
</div>
<div id="dependancy-installation" class="section level3">
<h3>Dependancy installation</h3>
<p><code>ascend</code> relies on packages found on CRAN and Bioconductor. Please install these packages before installing <code>ascend</code>.</p>
<div id="packages-from-cran" class="section level4">
<h4>1.3.1 Packages from CRAN</h4>
<p>You can use the install.packages() to install the packages described in this section. The pcakages you require from this repository are as follows:</p>
<ol style="list-style-type: decimal">
<li><a href="https://cran.r-project.org/web/packages/devtools/index.html">devtools</a>: This package will be used to load the development version of <code>ascend</code>.</li>
<li><a href="https://www.tidyverse.org/">tidyverse</a>: This is a series of R packages for data science and visualisation. This will install packages such as dplyr, ggplot2 and tidyr.</li>
<li><a href="https://github.com/Rdatatable/data.table/wiki/Installation">data.table</a>: Please follow the instructions on this page for your operating system.</li>
</ol>
<p>Remaining packages can be installed as follows:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="co"># List of packages to install</span></a>
<a class="sourceLine" id="cb1-2" data-line-number="2">cran_packages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;reshape2&quot;</span>, <span class="st">&quot;fields&quot;</span>, <span class="st">&quot;ggbeeswarm&quot;</span>, <span class="st">&quot;gridExtra&quot;</span>, </a>
<a class="sourceLine" id="cb1-3" data-line-number="3">                   <span class="st">&quot;dynamicTreeCut&quot;</span>, <span class="st">&quot;dendextend&quot;</span>, <span class="st">&quot;RColorBrewer&quot;</span>,</a>
<a class="sourceLine" id="cb1-4" data-line-number="4">                   <span class="st">&quot;locfit&quot;</span>, <span class="st">&quot;KernSmooth&quot;</span>)</a>
<a class="sourceLine" id="cb1-5" data-line-number="5"></a>
<a class="sourceLine" id="cb1-6" data-line-number="6"><span class="co"># Easy command to install all at once</span></a>
<a class="sourceLine" id="cb1-7" data-line-number="7"><span class="kw">install.packages</span>(cran_packages)</a></code></pre></div>
</div>
<div id="packages-from-bioconductor" class="section level4">
<h4>Packages from Bioconductor</h4>
<p>Bioconductor is a repository for R packages related to the analysis and comprehension of high-throughput genomic data. It uses a separate set of commands for the installation of packages.</p>
<div id="setting-up-bioconductor" class="section level5">
<h5>Setting up Bioconductor</h5>
<p>Use the following code to retrieve the latest installer from Bioconductor.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1">## try http:// if https:// URLs are not supported</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"><span class="kw">biocLite</span>()</a></code></pre></div>
<p>You can then install the Bioconductor packages using <code>biocLite</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1">bioconductor_packages &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Biobase&quot;</span>, <span class="st">&quot;BiocGenerics&quot;</span>, <span class="st">&quot;BiocParallel&quot;</span>,</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">                           <span class="st">&quot;SingleCellExperiment&quot;</span>, <span class="st">&quot;GenomeInfoDb&quot;</span>, <span class="st">&quot;GenomeInfoDbData&quot;</span>)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw">biocLite</span>(bioconductor_packages)</a></code></pre></div>
</div>
<div id="scaterscran-package-installation" class="section level5">
<h5>scater/scran package installation</h5>
<p><a href="https://bioconductor.org/packages/devel/bioc/html/scater.html">scater</a> and <a href="https://bioconductor.org/packages/devel/bioc/html/scran.html">scran</a> are scRNA-seq analysis toolboxes that provide more in-depth methods for QC and filtering. You may choose to install these packages if you wish to take advantage of the wrappers provided for these packages.</p>
</div>
<div id="differential-expression-packages" class="section level5">
<h5>Differential expression packages</h5>
<p><code>ascend</code> provides wrappers for <a href="https://bioconductor.org/packages/release/bioc/html/DESeq.html">DESeq</a> and <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a>, so you may choose to add them to your installation.</p>
</div>
<div id="possible-issues-with-the-stringi-package" class="section level5">
<h5>Possible issues with the ‘stringi’ package</h5>
<p>There may be issues for some users related to the R package “stringi”. This package is a dependancy for some of the packages from Bioconductor. Try installing this package from <a href="https://cran.r-project.org/web/packages/stringi/index.html%5D">this website</a></p>
</div>
</div>
</div>
</div>
<div id="installing-ascend-via-devtools" class="section level2">
<h2>Installing ‘ascend’ via devtools</h2>
<p>As <code>ascend</code> is still under development, we will use devtools to install the package.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Load devtools package</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2"><span class="kw">library</span>(devtools)</a>
<a class="sourceLine" id="cb4-3" data-line-number="3"></a>
<a class="sourceLine" id="cb4-4" data-line-number="4"><span class="co"># Use devtools to install the package</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="kw">install_github</span>(<span class="st">&quot;IMB-Computational-Genomics-Lab/ascend&quot;</span>, <span class="dt">ref =</span> <span class="st">&quot;devel&quot;</span>)</a>
<a class="sourceLine" id="cb4-6" data-line-number="6"><span class="co">#&gt; Downloading GitHub repo IMB-Computational-Genomics-Lab/ascend@devel</span></a>
<a class="sourceLine" id="cb4-7" data-line-number="7"><span class="co">#&gt; from URL https://api.github.com/repos/IMB-Computational-Genomics-Lab/ascend/zipball/devel</span></a>
<a class="sourceLine" id="cb4-8" data-line-number="8"><span class="co">#&gt; Installing ascend</span></a>
<a class="sourceLine" id="cb4-9" data-line-number="9"><span class="co">#&gt; '/Library/Frameworks/R.framework/Resources/bin/R' --no-site-file  \</span></a>
<a class="sourceLine" id="cb4-10" data-line-number="10"><span class="co">#&gt;   --no-environ --no-save --no-restore --quiet CMD INSTALL  \</span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="co">#&gt;   '/private/var/folders/zz/zyxvpxvq6csfxvn_n0001lp0000cxh/T/RtmpKsGqI4/devtools1b875f67fb0b/IMB-Computational-Genomics-Lab-ascend-0d98e33'  \</span></a>
<a class="sourceLine" id="cb4-12" data-line-number="12"><span class="co">#&gt;   --library='/Users/a.senabouth/Library/R/3.5/library' --install-tests</span></a>
<a class="sourceLine" id="cb4-13" data-line-number="13"><span class="co">#&gt; </span></a>
<a class="sourceLine" id="cb4-14" data-line-number="14"><span class="co">#&gt; Reloading installed ascend</span></a>
<a class="sourceLine" id="cb4-15" data-line-number="15"><span class="co">#&gt; Warning: replacing previous import 'BiocGenerics::Position' by</span></a>
<a class="sourceLine" id="cb4-16" data-line-number="16"><span class="co">#&gt; 'ggplot2::Position' when loading 'ascend'</span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="co">#&gt; Warning: replacing previous import 'S4Vectors::rowSums' by</span></a>
<a class="sourceLine" id="cb4-18" data-line-number="18"><span class="co">#&gt; 'Matrix::rowSums' when loading 'ascend'</span></a>
<a class="sourceLine" id="cb4-19" data-line-number="19"><span class="co">#&gt; Warning: replacing previous import 'S4Vectors::intersect' by</span></a>
<a class="sourceLine" id="cb4-20" data-line-number="20"><span class="co">#&gt; 'dplyr::intersect' when loading 'ascend'</span></a></code></pre></div>
<p>The package can then be loaded as normal.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Load the package in R</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw">library</span>(ascend)</a></code></pre></div>
</div>
<div id="configuring-biocparallel" class="section level2">
<h2>Configuring BiocParallel</h2>
<p>This package makes extensive use of <a href="http://bioconductor.org/packages/release/bioc/html/BiocParallel.html">BiocParallel</a>, enabling <code>ascend</code> to make the most of your computer’s hardware. As each system is different, BiocParallel needs to be configured by the user. Here are some example configurations.</p>
<div id="unixlinuxmacos-single-machine" class="section level3">
<h3>Unix/Linux/MacOS (Single Machine)</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">library</span>(BiocParallel)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">ncores &lt;-<span class="st"> </span>parallel<span class="op">::</span><span class="kw">detectCores</span>() <span class="op">-</span><span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb6-3" data-line-number="3"><span class="kw">register</span>(<span class="kw">MulticoreParam</span>(<span class="dt">workers =</span> ncores, <span class="dt">progressbar=</span><span class="ot">TRUE</span>), <span class="dt">default =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="windows-single-machine---quad-core-system" class="section level3">
<h3>Windows (Single Machine - Quad-core system)</h3>
<p>The following commands allows Windows to parallelise functions via BiocParallel. Unlike multicore processing in *nix systems, Snow creates additional R sessions to export tasks to. This requires additional computational resources to run and manage the tasks.</p>
<p>We recomend you bypass this step if your machine has lower specs.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1"><span class="kw">library</span>(BiocParallel)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">workers &lt;-<span class="st"> </span><span class="dv">3</span> <span class="co"># Number of cores on your machine - 1</span></a>
<a class="sourceLine" id="cb7-3" data-line-number="3"><span class="kw">register</span>(<span class="kw">SnowParam</span>(<span class="dt">workers =</span> workers, </a>
<a class="sourceLine" id="cb7-4" data-line-number="4">                   <span class="dt">type =</span> <span class="st">&quot;SOCK&quot;</span>, </a>
<a class="sourceLine" id="cb7-5" data-line-number="5">                   <span class="dt">progressbar =</span> <span class="ot">TRUE</span>), <span class="dt">default =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
