#' GenerateControlMetrics
#'
#' Function called by \code{\link{GenerateMetrics}} to generate metrics for 
#' control-related statistics.
#'
#' @param x Name of a control listed in Controls.
#' @param expression.matrix Transcript counts stored in a sparse matrix object.
#' @param control.list A list of controls supplied by the user.
#' @param total.counts List of total counts generated by \code{\link{GenerateMetrics}}.
#' @return A list containing a list of values
#' @importFrom Matrix colSums
GenerateControlMetrics <- function(x, expression.matrix = NULL, control.list = NULL, total.counts = NULL) {
    control.group <- control.list[[x]]
    control.bool <- rownames(expression.matrix) %in% control.group
    control.transcript.counts <- expression.matrix[control.bool, ]
    control.transcript.total.counts <- Matrix::colSums(control.transcript.counts)
    control.pt.matrix <- (control.transcript.total.counts/total.counts) * 100
    named.list <- list(ControlTranscriptCounts = control.transcript.total.counts, PercentageTotalCounts = control.pt.matrix)
    output.list <- list()
    output.list[[x]] <- named.list
    return(output.list)
}

#' GenerateMetrics
#' 
#' Generic method for \code{\link{GenerateMetrics}}.
#' 
#' @param x An EMSet
#' @return A list of metrics attached to \code{x}
#' 
#' @importFrom methods setGeneric
setGeneric(name = "GenerateMetrics", def = function(x) {
    standardGeneric("GenerateMetrics")
})

#' GenerateMetrics
#'
#' This function generates the following set of values per cell:
#' \itemize{
#' \item{\strong{Total Counts}: Total number of expressed transcripts per cell.}
#' \item{\strong{Total Feature Counts per Cell}: Number of non-control genes expressed per cell.}
#' \item{\strong{Total Expression}: Total number of expressed transcripts in the dataset.}
#' \item{\strong{Top Gene List}: Entire list of genes organised from most expressed to least expressed across the entire dataset.}
#' \item{\strong{Control Transcript Counts}: Total sum of control transcript counts per cell.}
#' \item{\strong{Percentage Total Counts}: Percentage of transcripts originating from control genes per cell.}
#' \item{\strong{AverageCounts}: Average transcript count for a gene.}
#' \item{\strong{GenesPerCell}: Number of unique transcripts expressed by a cell.}
#' \item{\strong{CellsPerGene}: Number of cells expressing a gene.}
#' \item{\strong{CountsPerGene}: Total number of transcripts produced by that gene expressed by all cells.}
#' \item{\strong{Mean Gene Expression}: Mean expression level of a gene across the entire dataset.}
#' }
#'
#' This function is called by \code{\link{NewEMSet}} and generates metrics for 
#' the new expression matrix. This function can also be called independantly, to 
#' update the metrics for a \code{\linkS4class{EMSet}} object.
#' 
#' @param x An EMSet
#' @include ascend_objects.R
#' @return A list of metrics attached to \code{x}
#' 
#' @importFrom BiocParallel bplapply
#' @importFrom Matrix colSums rowSums rowMeans
#' @importFrom methods setGeneric setMethod
#' 
setMethod("GenerateMetrics", signature("EMSet"), function(x) {
    # Retrieve required objects from EMSet
    expression.matrix <- x@ExpressionMatrix
    control.list <- x@Controls
    metrics.list <- list()

    ### Calculate library size per cell
    total.counts <- Matrix::colSums(expression.matrix)

    # User may have not supplied controls
    if (length(control.list) > 0) {
        # Prepare outputs
        control.transcript.counts.list <- list()
        percentage.lists.counts.list <- list()

        # Generate Metrics
        print("Calculating control metrics...")
        control.counts <- BiocParallel::bplapply(names(control.list), GenerateControlMetrics, expression.matrix = expression.matrix, control.list = control.list,
            total.counts = total.counts)

        # Unpackage stats
        unpacked.control.counts <- unlist(control.counts, recursive = FALSE)
        percentage.lists.counts.list <- lapply(names(unpacked.control.counts), function(x) unpacked.control.counts[[x]][["PercentageTotalCounts"]])
        names(percentage.lists.counts.list) <- names(unpacked.control.counts)
        control.transcript.counts.list <- lapply(names(unpacked.control.counts), function(x) unpacked.control.counts[[x]][["ControlTranscriptCounts"]])
        names(control.transcript.counts.list) <- names(unpacked.control.counts)

        # Calculate feature counts (Exclude controls)
        control.bool <- rownames(expression.matrix) %in% (unlist(control.list, use.names = FALSE))

        if (any(control.bool)) {
            endogenous.exprs.mtx <- expression.matrix[!control.bool, ]
        } else {
            endogenous.exprs.mtx <- expression.matrix
            percentage.lists.counts.list <- lapply(percentage.lists.counts.list, function(x) ifelse(is.na(x), 0, x))
            control.transcript.counts.list <- lapply(control.transcript.counts.list, function(x) ifelse(is.na(x), 0, x))
        }

        total.features.counts.per.cell <- Matrix::colSums(endogenous.exprs.mtx != 0)
        metrics.list <- c(metrics.list, list(TotalFeatureCountsPerCell = total.features.counts.per.cell, ControlTranscriptCounts = control.transcript.counts.list,
            PercentageTotalCounts = percentage.lists.counts.list))
    } else {
        x@Log$Controls <- FALSE
    }

    counts.per.gene <- Matrix::rowSums(expression.matrix)
    average.counts <- Matrix::rowMeans(expression.matrix)
    genes.per.cell <- Matrix::colSums(expression.matrix != 0)
    cells.per.gene <- Matrix::rowSums(expression.matrix != 0)
    mean.gene.expression <- Matrix::rowMeans(expression.matrix)

    # Calculate top gene expression
    total.expression <- sum(expression.matrix)
    sorted.counts.per.gene <- sort(counts.per.gene, decreasing = TRUE)
    top.gene.list <- names(sorted.counts.per.gene)
    top.gene.bool <- rownames(expression.matrix) %in% top.gene.list
    sorted.exprs.mtx <- expression.matrix[top.gene.bool, ]
    top.genes.percentage <- 100 * sum(sorted.counts.per.gene)/total.expression

    # Load generated values into the Metrics slot
    metrics.list <- c(metrics.list, list(TotalCounts = total.counts, TotalExpression = total.expression, TopGeneList = top.gene.list, TopGenesPercentage = top.genes.percentage,
        AverageCounts = average.counts, GenesPerCell = genes.per.cell, CellsPerGene = cells.per.gene, CountsPerGene = counts.per.gene, MeanGeneExpression = mean.gene.expression))
    remove(expression.matrix)
    x@Metrics <- metrics.list
    return(x)
})

#' AddControlInfo
#' Called by ascend object creation. Adds the control information to the dataframe.
#' This is a hidden function and shouldn't be used directly by users.
#' 
#' @param gene.information A dataframe containing gene identiers in the first 
#' column.
#' @param controls List of controls supplied by the user. They must match the
#' gene identifiers used in the expression matrix.
#' @return A dataframe (\code{gene.information}) that specifies if a gene is a 
#' \code{control}.
#' 
AddControlInfo <- function(gene.information, controls) {
    # Verify controls are in the cell information data frame
    gene.information$control <- rep(FALSE, nrow(gene.information))
    gene.information[gene.information[, 1] %in% unlist(controls), ]$control <- TRUE
    return(gene.information)
}

#' ConvertGeneAnnotation
#' 
#' Generic method for \code{\link{ConvertGeneAnnotation}}.
#' @param x A \code{\linkS4class{EMSet}} object.
#' @param old.annotation Name of the column containing the current gene annotations.
#' @param new.annotation Name of the column you would like to convert the gene 
#' 
#' @return An EMSet with the row names in the expression matrix switched to the
#' values in the new annotation column, and controls converted to their 
#' corresponding identifier in the new annotation.
#'
#' @examples
#' # Load EMSet that hasn't been filtered or normalised
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleEMSet.rds"))
#' 
#' # Convert annotation to ENSEMBL
#' EMSet <- ConvertGeneAnnotation(EMSet, "gene_id", "ensembl_id")
#' 
#' @importFrom methods setGeneric
#' @export
#' 
setGeneric(name = "ConvertGeneAnnotation", def = function(x, old.annotation, new.annotation) {
    standardGeneric("ConvertGeneAnnotation")
})

#' ConvertGeneAnnotation
#'
#' Convert gene identifiers used in this \code{\linkS4class{EMSet}} to 
#' identifiers used in another column of a dataframe stored in the 
#' GeneAnnotation slot.
#'
#' @param x A \code{\linkS4class{EMSet}} object.
#' @param old.annotation Name of the column containing the current gene annotations.
#' @param new.annotation Name of the column you would like to convert the gene 
#' annotations to.
#' 
#' @return An EMSet with the row names in the expression matrix switched to the
#' values in the new annotation column, and controls converted to their 
#' corresponding identifier in the new annotation.
#' 
#' @examples
#' # Load EMSet that hasn't been filtered or normalised
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleEMSet.rds"))
#' 
#' # Convert annotation to ENSEMBL
#' EMSet <- ConvertGeneAnnotation(EMSet, "gene_id", "ensembl_id")
#' 
#' @importFrom dplyr select everything
#' @importFrom methods setGeneric setMethod
#' @include ascend_objects.R
#' @export
#' 
setMethod("ConvertGeneAnnotation", signature("EMSet"), function(x, old.annotation, new.annotation) {
    # Get currently-used gene identifiers Load Gene Annotation
    gene.annotation <- x@GeneInformation

    ## From expression matrix
    present.rownames <- rownames(x@ExpressionMatrix)

    ## From control lists
    control.list <- x@Controls

    # Retrieve new values
    new.identifiers <- gene.annotation[, new.annotation]

    ## Rename expression matrix
    rownames(x@ExpressionMatrix) <- new.identifiers

    # Convert control list
    if (length(control.list) > 0) {
      updated.control.list <- sapply(names(control.list), function(control.name){
        gene.annotation[, new.annotation][which(gene.annotation[, old.annotation] 
                                                %in% control.list[[control.name]])]}, 
                                                USE.NAMES = TRUE)
      x@Controls <- updated.control.list
    }

    # Move new annotation to column 1
    updated.gene.info <- gene.annotation %>% dplyr::select(new.annotation, dplyr::everything())
    x@GeneInformation <- updated.gene.info
    
    # Regenerate metrics
    x <- GenerateMetrics(x)
    
    return(x)
})

#' ExcludeControl
#' 
#' Generic method for \code{\link{ExcludeControl}}.
#' @param x An \code{\linkS4class{EMSet}} object. It is recommended that 
#' you run this step after this object has undergone filtering.
#' @param control.name Name of the control set you want to remove from the dataset.
#' @return \code{x} without the \code{control.name} set of controls in the 
#' expression matrix.
#' 
#' @examples
#' # Load example EMSet
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleEMSet.rds"))
#' 
#' # Exclude "Mt" group of controls
#' EMSet <- ExcludeControl(EMSet, "Mt")
#' 
#' @importFrom methods setGeneric
#' @export
#' 
setGeneric(name = "ExcludeControl", def = function(x, control.name) {
    standardGeneric("ExcludeControl")
})

#' ExcludeControl
#'
#' Removes the specified control from the expression matrix. This should be used
#' after normalisation, and prior to PCA reduction, clustering and differential
#' expression analysis.
#' 
#' @param x An \code{\linkS4class{EMSet}} object. It is recommended that 
#' you run this step after this object has undergone filtering.
#' @param control.name Name of the control set you want to remove from the 
#' dataset.
#' @return \code{x} without the \code{control.name} set of controls in the 
#' expression matrix.
#' 
#' @examples
#' # Load example EMSet
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleEMSet.rds"))
#' 
#' # Exclude "Mt" group of controls
#' EMSet <- ExcludeControl(EMSet, "Mt")
#' 
#' @importFrom methods setGeneric setMethod
#' @export
#'
setMethod("ExcludeControl", signature("EMSet"), function(x, control.name) {
    # Identify indices of control genes in the current expression matrix 
    # Keep this matrix sparse for faster processing power
    expression.matrix <- x@ExpressionMatrix
    control.list <- x@Controls

    # We can sync up the slots once we update the expression matrix 
    # Convert the control list into a boolean so we can remove rows from the
    # sparse matrix
    control.list <- x@Controls[[control.name]]
    control.in.mtx <- rownames(expression.matrix) %in% control.list

    # Remove control genes from the matrix by identified matrices
    endogenous.exprs.mtx <- expression.matrix[!control.in.mtx, ]

    # Reload the expression matrix with the updated matrix
    x@ExpressionMatrix <- endogenous.exprs.mtx
    x <- SyncSlots(x)

    # Update the log
    updated.log <- list()
    updated.log[[control.name]] <- TRUE

    # Update the controls
    x@Controls[[control.name]] <- NULL

    # Update ExcludeControls
    if (is.null(x@Log$ExcludeControls)) {
        remove.log <- updated.log
        x@Log$ExcludeControls <- remove.log
    } else {
        x@Log$ExcludeControls <- c(x@Log$ExcludeControls, updated.log)
    }

    # Update Controls
    if (length(x@Controls) == 0) {
        x@Log$Controls <- FALSE
    }

    # Add removed genes to a list of removed genes
    if (is.null(x@Log$RemovedGenes)) {
        x@Log$RemovedGenes <- control.list
    } else {
        x@Log$RemovedGenes <- c(x@Log$RemovedGenes, control.list)
    }
    # Regenerate metrics and return the updated object
    x <- GenerateMetrics(x)
    return(x)
})

#' DisplayLog
#' 
#' Print out the log of an \code{\linkS4class{EMSet}}.
#' @param x An \linkS4class{EMSet}
#' @include ascend_objects.R
#' @return A print-out of the log attached to an \code{x}
#' 
#' @examples
#' # Load example EMSet
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleClusteredEMSet.rds"))
#' 
#' DisplayLog(EMSet)
#' 
#' @importFrom methods setGeneric
#' @export
setGeneric(name = "DisplayLog", def = function(x) {
    standardGeneric("DisplayLog")
})

#' DisplayLog
#'
#' Print out the log of an \code{\linkS4class{EMSet}}.
#' @param x An \linkS4class{EMSet}
#' @include ascend_objects.R
#' @return A print-out of the log attached to an \code{x}
#' 
#' @examples
#' # Load example EMSet
#' EMSet <- readRDS(system.file(package = "ascend", "extdata", 
#' "ExampleClusteredEMSet.rds"))
#' 
#' DisplayLog(EMSet)
#' 
#' @importFrom methods setGeneric setMethod
#' @export
setMethod("DisplayLog", signature("EMSet"), function(x) {
    log <- x@Log
    print(log)
})
